src/json/json.cc

```diff
@@ -258,6 +258,113 @@ STATIC JsonUtilCode parseSetCmdArgs(ValkeyModuleString **argv, const int argc, S
+        MSetCmdArgs &current_arg = args_list.get()[i];
+
+        current_arg.key_str = argv[i * 3 + 1];
+        current_arg.key = static_cast<ValkeyModuleKey*>(
+            ValkeyModule_OpenKey(ctx, current_arg.key_str, VALKEYMODULE_READ | VALKEYMODULE_WRITE));
+        current_arg.path = ValkeyModule_StringPtrLen(argv[i * 3 + 2], nullptr);
+        current_arg.json = ValkeyModule_StringPtrLen(argv[i * 3 + 3], &current_arg.json_len);
+
+        // Validate key type
+        int type = ValkeyModule_KeyType(current_arg.key);
+        if (type != VALKEYMODULE_KEYTYPE_EMPTY &&
+            ValkeyModule_ModuleTypeGetType(current_arg.key) != DocumentType) {
+            return JSONUTIL_NOT_A_DOCUMENT_KEY;
+        }
+
+        // Check if the key is new and validate root path requirement
+        bool is_new_valkey_key = (type == VALKEYMODULE_KEYTYPE_EMPTY);
+        current_arg.is_root_path = jsonutil_is_root_path(current_arg.path);
+
+        if (is_new_valkey_key && !current_arg.is_root_path) {
+            return JSONUTIL_COMMAND_SYNTAX_ERROR;
+        }
+
+        // Begin tracking memory
+        int64_t begin_val = jsonstats_begin_track_mem();
+
+        if (current_arg.is_root_path) {
+            JDocument *doc = nullptr;
+            rc = dom_parse(ctx, current_arg.json, current_arg.json_len, &doc);
+            if (rc != JSONUTIL_SUCCESS) {
+                if (doc) dom_free_doc(doc);
+                return rc;
+            }
+
+            // End tracking memory and compute document size
+            int64_t delta = jsonstats_end_track_mem(begin_val);
+            current_arg.doc_size = dom_get_doc_size(doc) + delta;
+
+            if (json_is_instrument_enabled_insert() || json_is_instrument_enabled_update()) {
+                size_t len;
+                const char* key_cstr = ValkeyModule_StringPtrLen(current_arg.key_str, &len);
+                std::size_t key_hash = std::hash<std::string_view>{}(std::string_view(key_cstr, len));
+                ValkeyModule_Log(ctx, "warning",
+                                "Dump document structure before setting JSON key (hashed) %zu whole doc %p:",
+                                key_hash, static_cast<void*>(doc));
+                DumpRedactedJValue(doc->GetJValue(), nullptr, "warning");
+            }
+
+            dom_free_doc(doc);
+        } else {
+            JDocument *doc = static_cast<JDocument*>(ValkeyModule_ModuleTypeGetValue(current_arg.key));
+            if (doc == nullptr) {
+                return JSONUTIL_DOCUMENT_KEY_NOT_FOUND;
+            }
+
+            rc = dom_verify_value(ctx, doc, current_arg.path, current_arg.json);
+            if (rc != JSONUTIL_SUCCESS) {
+                dom_free_doc(doc);
+                return rc;
+            }
+
+            // End tracking memory and compute document size
+            int64_t delta = jsonstats_end_track_mem(begin_val);
+            current_arg.doc_size = dom_get_doc_size(doc) + delta;
+            dom_free_doc(doc);
```

Comment:
Memory corruption! We should not free the doc here because the doc was just fetched from the dict. We didn't create it here.